src/native/corehost/hostmisc/pal.windows.cpp

```diff
@@ -727,8 +727,91 @@ bool pal::clr_palstring(const char* cstr, pal::string_t* out)
     return wchar_convert_helper(CP_UTF8, cstr, ::strlen(cstr), out);
 }
 
-// Return if path is valid and file exists, return true and adjust path as appropriate.
-bool pal::realpath(string_t* path, bool skip_error_logging)
+typedef std::unique_ptr<std::remove_pointer<HANDLE>::type, decltype(&::CloseHandle)> SmartHandle;
+
+// Like fullpath, but resolves symlinks.
+bool pal::realpath(pal::string_t* path, bool skip_error_logging)
+{
+    if (path->empty())
+    {
+        return false;
+    }
+
+    // Use CreateFileW + GetFinalPathNameByHandleW to resolve symlinks
+
+    SmartHandle file(
+        ::CreateFileW(
+            path->c_str(),
+            0, // Querying only
+            FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+            nullptr, // default security
+            OPEN_EXISTING, // existing file
+            FILE_ATTRIBUTE_NORMAL, // normal file
+            nullptr), // No attribute template
+        &::CloseHandle);
+
+    pal::char_t buf[MAX_PATH];
+    size_t size;
+
+    if (file.get() == INVALID_HANDLE_VALUE)
+    {
+        // If we get "access denied" that may mean the path represents a directory.
```

Comment:
Is this an intentional limitation? `CreateFile` should be able to get a directory handle with `FILE_FLAG_BACKUP_SEMANTICS`.

Reply:
To be honest I can't remember. There seem to be a lot of caveats in `FILE_FLAG_BACKUP_SEMANTICS` docs. Since we're going to fullpath in this case anyway, is there any point in proceeding?

Also, does Windows have directory symlinks?

Reply:
Added a comment. The file security implications of `FILE_FLAG_BACKUP_SEMANTICS` worry me a bit just because I don't understand all the consequences.